### PR TITLE
Harden result source handling

### PR DESCRIPTION
**Pull request by EasyRhinoMSFT**
Correct handling when no source is found
Improve ResultSourceHost functional pattern
Add unit test

### CODE REVIEW

**Comment by yongyan-gh** [comment]
>             foreach (Type type in resultSourceTypes)

nit: this.resultSourceTypes

---
Refers to: src/Sarif.Viewer.VisualStudio.ResultSources.Factory.Core/ResultSourceFactory.cs:76 in de64bb0. [](commit_id = de64bb0bd38eb2f1e0d0ef19dbae48ede612c3cf, deletion_comment = False)

File: src/Sarif.Viewer.VisualStudio.Core/SarifViewerPackage.cs [hunk 1]
```diff
@@ -171,9 +171,10 @@ private async Task InitializeResultSourceHostAsync()
             string solutionPath = GetSolutionDirectoryPath();
             if (!string.IsNullOrWhiteSpace(solutionPath) && SarifViewerOption.Instance.IsGitHubAdvancedSecurityEnabled)
             {
-                this.resultSourceHost = new ResultSourceHost(solutionPath, this);
+                var resultSourceFactory = new ResultSourceFactory(solutionPath, this);
+                this.resultSourceHost = new ResultSourceHost();
                 this.resultSourceHost.ResultsUpdated += this.ResultSourceHost_ResultsUpdated;
-                await this.resultSourceHost.RequestAnalysisResultsAsync();
+                await this.resultSourceHost.RequestAnalysisResultsAsync(resultSourceFactory);
```
Comment (yongyan-gh) [inline src/Sarif.Viewer.VisualStudio.Core/SarifViewerPackage.cs:177]: >RequestAnalysisResultsAsync [](http://example.com/codeflow?start=44&length=27)

is it possible this call or the initializing of ResultSourceFactory/Host throw exception? should we catch here? Just want to make sure if any unexpected error happens inside ResultsourceFactory/Host, rest of the viewer extension can still run.

File: src/Sarif.Viewer.VisualStudio.Core/SarifViewerPackage.cs [hunk 2]
```diff
@@ -171,9 +171,10 @@ private async Task InitializeResultSourceHostAsync()
             string solutionPath = GetSolutionDirectoryPath();
             if (!string.IsNullOrWhiteSpace(solutionPath) && SarifViewerOption.Instance.IsGitHubAdvancedSecurityEnabled)
             {
-                this.resultSourceHost = new ResultSourceHost(solutionPath, this);
+                var resultSourceFactory = new ResultSourceFactory(solutionPath, this);
```
Comment (yongyan-gh) [inline src/Sarif.Viewer.VisualStudio.Core/SarifViewerPackage.cs:174]: >solutionPath [](http://example.com/codeflow?start=66&length=12)

In some cases the solution is not placed in the repo root path, its placed in a sub directory of repo root. E.g. {RepoRootPath}\src, does it work in this case?

Comment (EasyRhinoMSFT) [reply src/Sarif.Viewer.VisualStudio.Core/SarifViewerPackage.cs:174]: Yes it does. (It's the result source's responsibility.)

**Pull request merged**